Track persistent obstacles between frames and update pipeline to utilize ObstacleTracker

mrouie opened the pull request:
This pull request uses ObstacleTracker to track obstacles between frames and updates the pipeline to utilize ObstacleTracker functions

ObstacleTracker uses ObstacleTracker.update() to update the internal buffer of boundingboxes in the current frame with a 10% (radius) movement tolerance per obstacle. 
It also uses ObstacleTracker.getPersistentObstacles() to return a list of BoundingBoxes that contain obstacles that have been persistent for 5 frames.

mrouie committed:
Track persistent obstacles between frames and update pipeline to utilize ObstacleTracker

fallscameron01 commented on vision/obstacle/obstacle_tracker.py:
```diff
@@ -0,0 +1,95 @@
+        obj1: Obstacle
+            The obstacle of reference
+        obj2: Obstacle
+            The obstacle in question
```
Return in documentation

fallscameron01 commented on vision/obstacle/obstacle_tracker.py:
```diff
@@ -0,0 +1,95 @@
+        obj2: Obstacle
+            The obstacle in question
+        """
+        return obj1.center >= (1-self.MVMT_TOLERANCE)*obj2.center and obj1.center <= (1+self.MVMT_TOLERANCE)*obj2.center
```
Looks a little messy, should put parenthesis around the statements with the and in between:
`(obj1.center >= (1-self.MVMT_TOLERANCE)*obj2.center) and (obj1.center <= (1+self.MVMT_TOLERANCE)*obj2.center)`

fallscameron01 replied:
Since the center has both an x and a y value, you need to compare x and y separately, then logical and the results from each.

fallscameron01 commented on vision/obstacle/obstacle_tracker.py:
```diff
@@ -0,0 +1,95 @@
+            return
+
+        # create an array of obstacles that existed in the previous (buffer) frame and replace the buffer with it
+        for i in np.arange(new_obstacles.size()):
```
`new_obstacles.size()`

Size is actually a variable of an ndarray, not a function. No parenthesis needed.

fallscameron01 commented on vision/obstacle/obstacle_tracker.py:
```diff
@@ -0,0 +1,95 @@
+
+        # create an array of obstacles that existed in the previous (buffer) frame and replace the buffer with it
+        for i in np.arange(new_obstacles.size()):
+            if(_isSameObstacle(new_obstacles[i], self.obstacles[i])):
```
Should be

`self._isSameObstacle(`

since _isSameObstacle is a member function.

fallscameron01 commented on vision/obstacle/obstacle_tracker.py:
```diff
@@ -0,0 +1,95 @@
+        list[BoundingBox]
+            a list of the persistent obstacle BoundingBoxes
+        """
+        persistent_obstacles = np.array([])
```
Your return value should be a normal Python list since that is how bounding boxes are stored in the pipeline.

fallscameron01 commented on vision/obstacle/obstacle_tracker.py:
```diff
@@ -0,0 +1,95 @@
+        persistent_obstacles = np.array([])
+
+        # append only obstacles persistent for PERSISTENCE_THRESHOLD frames
+        for i in np.arange(self.obstacles.size()):
```
Same issue above, size should not have parenthesis.

fallscameron01 commented on vision/obstacle/obstacle_tracker.py:
```diff
@@ -0,0 +1,95 @@
+    """
+    def __init__(self, bbox: BoundingBox):
+        self.bounding_box = bbox
+        self.center = np.mean(bbox)
```
`np.mean(bbox)` will not work because bbox is a user created class. You need to take the mean of bbox.vertices
Also, we want to separate the x and y values, so we should take the mean along the 0th axis.

`np.mean(bbox.vertices, axis=0)` will return [xval, yval] with a datatype of ndarray and datatypes of x and y as floats.

fallscameron01 commented on vision/obstacle/obstacle_tracker.py:
```diff
@@ -0,0 +1,95 @@
+        obj2: Obstacle
+            The obstacle in question
+        """
+        return obj1.center >= (1-self.MVMT_TOLERANCE)*obj2.center and obj1.center <= (1+self.MVMT_TOLERANCE)*obj2.center
```
One more thing with the math here: You are multiplying the center value by the movement tolerance, but you want to be multiplying by the total dimensions of the image. Otherwise, obstacles closer to the top left corner will have a smaller tolerance than obstacles in the lower right.

fallscameron01 replied:
Second sentence is misworded, should read:
You are multiplying the center value by the movement tolerance, but you want to be multiplying the total dimensions of the image by the movement tolerance.

Also, the image dimensions are 1920x1080.

fallscameron01 reviewed (changes requested):
Overall, the structure looks really good, but some subtleties are missing here. The center math and the math that compares the centers needs to be changed. Also there are several things that would give runtime errors, such as calling numpy's size as a function.
Pipeline was updated correctly though.

mrouie committed:
Resolve errors and rework calculation method for obstacle tracking

- Obstacle tracking now uses definite pixels as a tolerance for movement instead of a position-based percentage

fallscameron01 reviewed (approved):
Looks good. Ready to merge.
